File: visdaxClient.js
/**
 * Visdax Web SDK v1.1.0
 * B2B High-Fidelity Vision Restoration Client
 * Optimized for both Single and Batch Operations
 */
class VisdaxClient {
    constructor(config) {
        this.baseUrl = "https://api.visdax.com/api/v1";
        this.apiKey = config.apiKey;
        this.project = config.project;
        this.bucket = config.bucket;
        
        this.cacheName = 'visdax-vault-v1';
        this.maxSizeBytes = 500 * 1024 * 1024; // 500MB Hard Limit
    }

    _getHeaders() {
        return {
            "Authorization": `Bearer ${this.apiKey}`,
            "X-Visdax-Project": this.project,
            "X-Visdax-Bucket": this.bucket,
            "Content-Type": "application/json"
        };
    }

    // ==========================================
    // 1. SINGLE & BATCH SUBMISSION (UPLOAD)
    // ==========================================

    /**
     * Uploads a single file.
     * @param {File} file 
     */
    async submit(file) {
        const formData = new FormData();
        formData.append('file', file);
        
        const resp = await fetch(`${this.baseUrl}/post_file`, {
            method: 'POST',
            headers: { "Authorization": `Bearer ${this.apiKey}` },
            body: formData
        });
        
        if (!resp.ok) throw new Error(`Visdax: Upload Failed (${resp.status})`);
        return await resp.json();
    }

    /**
     * Uploads multiple files in sequence.
     * @param {FileList|File[]} files 
     */
    async submitBatch(files) {
        const results = [];
        for (const file of files) {
            try {
                const res = await this.submit(file);
                results.push(res);
            } catch (err) {
                results.push({ file: file.name, ok: false, error: err.message });
            }
        }
        return results;
    }

    // ==========================================
    // 2. SINGLE & BATCH RETRIEVAL (DOWNLOAD + LRU)
    // ==========================================

    /**
     * Loads a single asset.
     * Wraps loadBatch for consistent logic and caching.
     */
    async load(key) {
        const result = await this.loadBatch([key]);
        return result[key];
    }

    /**
     * Validates and loads high-fidelity images into the 500MB vault.
     * @param {string[]} keys - List of asset keys.
     * @returns {Object} Map of { key: blobUrl }
     */
    async loadBatch(keys) {
        const cache = await caches.open(this.cacheName);
        const etags = {};

        // 1. Check local vault for existing assets
        for (const key of keys) {
            const cached = await cache.match(key);
            if (cached) etags[key] = cached.headers.get('ETag');
        }

        // 2. Single 'Gatekeeper' revalidation ping
        const response = await fetch(`${this.baseUrl}/get_multifiles?restore=true`, {
            method: 'POST',
            headers: this._getHeaders(),
            body: JSON.stringify({ keys, etags })
        });

        if (!response.ok) throw new Error(`Visdax: API Error ${response.status}`);

        const data = await response.json();
        const finalUrls = {};

        // 3. Process results
        for (const asset of data.assets) {
            const key = asset.key;

            if (asset.status === 304) {
                // Verified active sub: Load from local cache
                const cachedResp = await cache.match(key);
                finalUrls[key] = URL.createObjectURL(await cachedResp.blob());
            } 
            else if (asset.status === 200) {
                // New/Updated content: Stream to local vault
                const blob = await (await fetch(`data:image/webp;base64,${asset.content}`)).blob();
                await this._enforceLRU(blob.size);
                
                const mockResponse = new Response(blob, {
                    headers: { 
                        'ETag': asset.etag, 
                        'Date': new Date().toUTCString(),
                        'Content-Type': 'image/webp'
                    }
                });
                
                await cache.put(key, mockResponse);
                finalUrls[key] = URL.createObjectURL(blob);
            }
        }
        return finalUrls;
    }

    async _enforceLRU(newSize) {
        const cache = await caches.open(this.cacheName);
        const keys = await cache.keys();
        let currentSize = 0;
        
        const entries = [];
        for (const request of keys) {
            const res = await cache.match(request);
            const blob = await res.blob();
            currentSize += blob.size;
            entries.push({ request, size: blob.size });
        }

        let i = 0;
        while (currentSize + newSize > this.maxSizeBytes && i < entries.length) {
            const oldest = entries[i];
            currentSize -= oldest.size;
            await cache.delete(oldest.request);
            i++;
        }
    }
}

File: visdaxClient.py
import os, requests, hashlib, base64, shutil, time
from pathlib import Path
from pqdm.threads import pqdm

class VisdaxClient:
    def __init__(self, api_key, project, bucket, limit_mb=500):
        self.api_key = api_key
        self.project = project
        self.bucket = bucket
        self.limit = limit_mb * 1024 * 1024
        self.cache_path = Path("~/.visdax_cache").expanduser()
        self.cache_path.mkdir(parents=True, exist_ok=True)
        self.base_url = "https://api.visdax.com/api/v1"
        
        # Reads the secret directly from the environment where the SDK is running
        

    def _get_headers(self):
        return {
            "Authorization": f"Bearer {self.api_key}",
            "X-Visdax-Project": self.project,
            "X-Visdax-Bucket": self.bucket,
        }

    # ==========================================
    # 1. SUBMISSION (UPLOAD) FUNCTIONS
    # ==========================================

    def submit(self, file_path):
        """Single file upload to 56k-Atomic Engine."""
        with open(file_path, 'rb') as f:
            files = {'file': (os.path.basename(file_path), f)}
            resp = requests.post(
                f"{self.base_url}/post_file", 
                headers=self._get_headers(), 
                files=files
            )
        return resp.json()

    def submit_batch(self, file_paths, n_jobs=4):
        """Parallel upload for large datasets."""
        return pqdm(file_paths, self.submit, n_jobs=n_jobs)

    # ==========================================
    # 2. RETRIEVAL (DOWNLOAD + LRU) FUNCTIONS
    # ==========================================

    def _enforce_lru(self, incoming_size):
        """Strictly keeps the cache folder under 500MB."""
        files = sorted(self.cache_path.glob("*.webp"), key=os.path.getmtime)
        current_size = sum(f.stat().st_size for f in files)
        while current_size + incoming_size > self.limit and files:
            oldest = files.pop(0)
            current_size -= oldest.stat().st_size
            oldest.unlink()

    def load(self, key):
        """Single asset load with secret verification."""
        return self.load_batch([key])[0]

    def load_batch(self, keys):
        """
        The Core ML Function: Validates a batch of ETags.
        Ensures the secret is passed to the server for 'Compulsory Cache' checks.
        """
        etags = {k: hashlib.md5(k.encode()).hexdigest() for k in keys}
        existing_etags = {k: v for k, v in etags.items() if (self.cache_path / f"{v}.webp").exists()}

        payload = {"keys": keys, "etags": existing_etags}
        
        resp = requests.post(
            f"{self.base_url}/get_multifiles?restore=true", 
            json=payload, 
            headers=self._get_headers()
        )
        
        if resp.status_code != 200:
            if resp.status_code == 403:
                raise Exception("Visdax Access Denied: Missing or invalid Internal SDK Secret.")
            raise Exception(f"Visdax Batch Failed: {resp.status_code} - {resp.text}")

        data = resp.json()
        final_paths = []

        for asset in data.get("assets", []):
            key = asset['key']
            local_file = self.cache_path / f"{etags[key]}.webp"

            if asset['status'] == 304:
                # Cache Hit: Refresh time for LRU
                os.utime(local_file, None)
                final_paths.append(str(local_file))
            
            elif asset['status'] == 200:
                # Cache Miss: Download new high-fidelity data
                content = base64.b64decode(asset['content'])
                self._enforce_lru(len(content))
                local_file.write_bytes(content)
                final_paths.append(str(local_file))
            
            else:
                print(f"Visdax Error: Asset {key} failed with status {asset['status']}")

        return final_paths
